File: src/lib/actions/reminders.ts
"use server";

import { and, eq } from "drizzle-orm";
import { revalidatePath } from "next/cache";
import { db } from "@/db";
import { reminders, users } from "@/db/schema";
import { requireActiveUser } from "@/lib/auth/current-user";
import type { FormState } from "./form-state";

function parseDateInput(value: string): Date | null {
  if (!value) return null;
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) return null;
  const date = new Date(`${value}T00:00:00`);
  return Number.isNaN(date.getTime()) ? null : date;
}

export async function createReminder(
  _prevState: FormState,
  formData: FormData,
): Promise<FormState> {
  const { household, user } = await requireActiveUser();

  const title = String(formData.get("title") ?? "").trim();
  const notes = String(formData.get("notes") ?? "").trim();
  const dueAt = parseDateInput(String(formData.get("dueAt") ?? ""));
  let assignedToUserId = String(formData.get("assignedToUserId") ?? "") || null;

  if (!title) return { error: "Title is required." };

  // Non-admins can only create reminders for themselves or the whole
  // household — not assign work to someone else.
  if (
    user.role !== "admin" &&
    assignedToUserId &&
    assignedToUserId !== user.id
  ) {
    assignedToUserId = user.id;
  }

  if (assignedToUserId) {
    const assignee = await db.query.users.findFirst({
      where: and(
        eq(users.id, assignedToUserId),
        eq(users.householdId, household.id),
      ),
    });
    if (!assignee) return { error: "Assignee not found." };
  }

  await db.insert(reminders).values({
    id: crypto.randomUUID(),
    householdId: household.id,
    title,
    notes: notes || null,
    dueAt,
    assignedToUserId,
  });

  revalidatePath("/reminders");
  revalidatePath("/");
  revalidatePath("/calendar");
  return null;
}

export async function toggleReminderDone(formData: FormData) {
  const { household } = await requireActiveUser();

  const id = String(formData.get("id") ?? "");
  const reminder = await db.query.reminders.findFirst({
    where: and(eq(reminders.id, id), eq(reminders.householdId, household.id)),
  });
  if (!reminder) return;

  await db
    .update(reminders)
    .set({ isDone: !reminder.isDone })
    .where(eq(reminders.id, id));

  revalidatePath("/reminders");
  revalidatePath("/");
  revalidatePath("/calendar");
}

export async function deleteReminder(formData: FormData) {
  const { household, user } = await requireActiveUser();

  const id = String(formData.get("id") ?? "");
  const reminder = await db.query.reminders.findFirst({
    where: and(eq(reminders.id, id), eq(reminders.householdId, household.id)),
  });
  if (!reminder) return;

  if (user.role !== "admin" && reminder.assignedToUserId !== user.id) {
    throw new Error("You can only delete your own reminders.");
  }

  await db.delete(reminders).where(eq(reminders.id, id));

  revalidatePath("/reminders");
  revalidatePath("/");
  revalidatePath("/calendar");
}

File: backend/internal/api/handlers/reminders.go
package handlers

import (
	"errors"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"

	"hestia/backend/internal/api/middleware"
	"hestia/backend/internal/services"
)

func (d *Deps) ListReminders(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	reminders, err := d.Reminder.List(household.ID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to list reminders"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"reminders": reminders})
}

type reminderRequest struct {
	Title            string  `json:"title"`
	Notes            *string `json:"notes"`
	DueAt            string  `json:"dueAt"` // YYYY-MM-DD, optional
	AssignedToUserID *string `json:"assignedToUserId"`
}

func (d *Deps) CreateReminder(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	user := middleware.CurrentUser(c)

	var req reminderRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request"})
		return
	}
	if req.Title == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "title is required"})
		return
	}

	var dueAt *time.Time
	if req.DueAt != "" {
		parsed, err := time.ParseInLocation("2006-01-02", req.DueAt, time.Local)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid due date"})
			return
		}
		dueAt = &parsed
	}

	// Non-HoHs can only create reminders for themselves or the whole
	// household — not assign work to someone else.
	assignedToUserID := req.AssignedToUserID
	if user.Role != "hoh" && assignedToUserID != nil && *assignedToUserID != user.ID {
		assignedToUserID = &user.ID
	}

	reminder, err := d.Reminder.Create(household.ID, services.ReminderInput{
		Title: req.Title, Notes: req.Notes, DueAt: dueAt, AssignedToUserID: assignedToUserID,
	})
	if err != nil {
		if errors.Is(err, services.ErrNotFound) {
			c.JSON(http.StatusBadRequest, gin.H{"error": "assignee not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to create reminder"})
		return
	}
	c.JSON(http.StatusCreated, reminder)
}

func (d *Deps) ToggleReminderDone(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	reminder, err := d.Reminder.ToggleDone(household.ID, c.Param("id"))
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "reminder not found"})
		return
	}
	c.JSON(http.StatusOK, reminder)
}

func (d *Deps) DeleteReminder(c *gin.Context) {
	household := middleware.CurrentHousehold(c)
	user := middleware.CurrentUser(c)

	err := d.Reminder.Delete(household.ID, c.Param("id"), user.ID, user.Role)
	if err != nil {
		switch {
		case errors.Is(err, services.ErrNotFound):
			c.JSON(http.StatusNotFound, gin.H{"error": "reminder not found"})
		case errors.Is(err, services.ErrForbidden):
			c.JSON(http.StatusForbidden, gin.H{"error": "you can only delete your own reminders"})
		default:
			c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to delete reminder"})
		}
		return
	}
	c.JSON(http.StatusOK, gin.H{"ok": true})
}
